### PR TITLE
upgrade to officer on addOfficer if already added

### LINKED ISSUE #473: addOfficer doesn't overwrite existing entry of that name
### Minecolonies version

minecolonies-1.10.2-0.6.2305

I added myself to a player's colony earlier tonight, using addOfficer. When I got sick of the builder-spam I went to remove myself but I clicked the demote button instead of delete.

re-running the addOfficer command gives the usual success message however the permission level isn't changed; I'm stuck at Friend.

Please have each run of addOfficer overwrite any existing perms for that username.

Cheers!

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/41090016-addofficer-doesn-t-overwrite-existing-entry-of-that-name?utm_campaign=plugin&utm_content=tracker%2F43175463&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F43175463&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #473 

# Changes proposed in this pull request:
- Should now allow to upgrade rank via addOfficer.
Still taking care to not overwrite owner.

Review please

### CODE REVIEW

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 33 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 30 minor

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Permissions.java#L325](https://github.com/Minecolonies/minecolonies/blob/8ff2366807f3eddf5f5140abc1c44b344cc68b01/src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java#L325): Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [Permissions.java#L779](https://github.com/Minecolonies/minecolonies/blob/8ff2366807f3eddf5f5140abc1c44b344cc68b01/src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java#L779): Remove this unused private "isSubscriber" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)
1. ![MAJOR][MAJOR] [Permissions.java#L864](https://github.com/Minecolonies/minecolonies/blob/8ff2366807f3eddf5f5140abc1c44b344cc68b01/src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java#L864): Reduce this class from 116 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MINOR][MINOR] [blockout](https://github.com/Minecolonies/minecolonies/blob/8ff2366807f3eddf5f5140abc1c44b344cc68b01/src/main/java/com/minecolonies/blockout): Add a 'package-info.java' file to document the 'blockout' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [controls](https://github.com/Minecolonies/minecolonies/blob/8ff2366807f3eddf5f5140abc1c44b344cc68b01/src/main/java/com/minecolonies/blockout/controls): Add a 'package-info.java' file to document the 'controls' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [views](https://github.com/Minecolonies/minecolonies/blob/8ff2366807f3eddf5f5140abc1c44b344cc68b01/src/main/java/com/minecolonies/blockout/views): Add a 'package-info.java' file to document the 'views' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [blocks](https://github.com/Minecolonies/minecolonies/blob/8ff2366807f3eddf5f5140abc1c44b344cc68b01/src/main/java/com/minecolonies/coremod/blocks): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [gui](https://github.com/Minecolonies/minecolonies/blob/8ff2366807f3eddf5f5140abc1c44b344cc68b01/src/main/java/com/minecolonies/coremod/client/gui): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [render](https://github.com/Minecolonies/minecolonies/blob/8ff2366807f3eddf5f5140abc1c44b344cc68b01/src/main/java/com/minecolonies/coremod/client/render): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [colony](https://github.com/Minecolonies/minecolonies/blob/8ff2366807f3eddf5f5140abc1c44b344cc68b01/src/main/java/com/minecolonies/coremod/colony): Add a 'package-info.java' file to document the 'colony' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 33 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 30 minor

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Permissions.java#L325](https://github.com/Minecolonies/minecolonies/blob/5ef95a20825ea76b103bdb8523b60bac45eff9dc/src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java#L325): Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [Permissions.java#L779](https://github.com/Minecolonies/minecolonies/blob/5ef95a20825ea76b103bdb8523b60bac45eff9dc/src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java#L779): Remove this unused private "isSubscriber" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)
1. ![MAJOR][MAJOR] [Permissions.java#L864](https://github.com/Minecolonies/minecolonies/blob/5ef95a20825ea76b103bdb8523b60bac45eff9dc/src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java#L864): Reduce this class from 116 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MINOR][MINOR] [blockout](https://github.com/Minecolonies/minecolonies/blob/5ef95a20825ea76b103bdb8523b60bac45eff9dc/src/main/java/com/minecolonies/blockout): Add a 'package-info.java' file to document the 'blockout' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [controls](https://github.com/Minecolonies/minecolonies/blob/5ef95a20825ea76b103bdb8523b60bac45eff9dc/src/main/java/com/minecolonies/blockout/controls): Add a 'package-info.java' file to document the 'controls' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [views](https://github.com/Minecolonies/minecolonies/blob/5ef95a20825ea76b103bdb8523b60bac45eff9dc/src/main/java/com/minecolonies/blockout/views): Add a 'package-info.java' file to document the 'views' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [blocks](https://github.com/Minecolonies/minecolonies/blob/5ef95a20825ea76b103bdb8523b60bac45eff9dc/src/main/java/com/minecolonies/coremod/blocks): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [gui](https://github.com/Minecolonies/minecolonies/blob/5ef95a20825ea76b103bdb8523b60bac45eff9dc/src/main/java/com/minecolonies/coremod/client/gui): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [render](https://github.com/Minecolonies/minecolonies/blob/5ef95a20825ea76b103bdb8523b60bac45eff9dc/src/main/java/com/minecolonies/coremod/client/render): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [colony](https://github.com/Minecolonies/minecolonies/blob/5ef95a20825ea76b103bdb8523b60bac45eff9dc/src/main/java/com/minecolonies/coremod/colony): Add a 'package-info.java' file to document the 'colony' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java [hunk 1]
```diff
@@ -718,7 +718,7 @@ public boolean addPlayer(@NotNull final String player, final Rank rank, final Wo
         }
         final GameProfile gameprofile = world.getMinecraftServer().getPlayerProfileCache().getGameProfileForUsername(player);
         //Check if the player already exists so that their rank isn't overridden
-        return gameprofile != null && !players.containsKey(gameprofile.getId()) && addPlayer(gameprofile, rank);
+        return gameprofile != null && !getOwner().equals(gameprofile.getId()) && addPlayer(gameprofile, rank);
```
Comment (cltnschlosser) [inline src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java]: Don't merge this yet, this creates a small security issue. I'll look for the best solution tonight.

Allows Officers to demote eachother.

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 39 issues
* ![MAJOR][MAJOR] 9 major
* ![MINOR][MINOR] 30 minor

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Permissions.java#L29](https://github.com/Minecolonies/minecolonies/blob/125975186ced3cf7687e445cea53ce34209efcb3/src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java#L29): class "Permissions" has 36 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [Permissions.java#L325](https://github.com/Minecolonies/minecolonies/blob/125975186ced3cf7687e445cea53ce34209efcb3/src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java#L325): Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [Permissions.java#L786](https://github.com/Minecolonies/minecolonies/blob/125975186ced3cf7687e445cea53ce34209efcb3/src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java#L786): Remove this unused private "isSubscriber" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)
1. ![MAJOR][MAJOR] [Permissions.java#L871](https://github.com/Minecolonies/minecolonies/blob/125975186ced3cf7687e445cea53ce34209efcb3/src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java#L871): Reduce this class from 116 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [PermissionsMessage.java#L48](https://github.com/Minecolonies/minecolonies/blob/125975186ced3cf7687e445cea53ce34209efcb3/src/main/java/com/minecolonies/coremod/network/messages/PermissionsMessage.java#L48): Reduce this class from 33 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [PermissionsMessage.java#L99](https://github.com/Minecolonies/minecolonies/blob/125975186ced3cf7687e445cea53ce34209efcb3/src/main/java/com/minecolonies/coremod/network/messages/PermissionsMessage.java#L99): Reduce this class from 63 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [PermissionsMessage.java#L185](https://github.com/Minecolonies/minecolonies/blob/125975186ced3cf7687e445cea53ce34209efcb3/src/main/java/com/minecolonies/coremod/network/messages/PermissionsMessage.java#L185): Reduce this class from 43 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [PermissionsMessage.java#L247](https://github.com/Minecolonies/minecolonies/blob/125975186ced3cf7687e445cea53ce34209efcb3/src/main/java/com/minecolonies/coremod/network/messages/PermissionsMessage.java#L247): Reduce this class from 59 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [PermissionsMessage.java#L331](https://github.com/Minecolonies/minecolonies/blob/125975186ced3cf7687e445cea53ce34209efcb3/src/main/java/com/minecolonies/coremod/network/messages/PermissionsMessage.java#L331): Reduce this class from 46 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MINOR][MINOR] [blockout](https://github.com/Minecolonies/minecolonies/blob/125975186ced3cf7687e445cea53ce34209efcb3/src/main/java/com/minecolonies/blockout): Add a 'package-info.java' file to document the 'blockout' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 39 issues
* ![MAJOR][MAJOR] 9 major
* ![MINOR][MINOR] 30 minor

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Permissions.java#L29](https://github.com/Minecolonies/minecolonies/blob/1c86e4ce5bb6d3ca11b572017dcfd315270f4465/src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java#L29): class "Permissions" has 36 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [Permissions.java#L325](https://github.com/Minecolonies/minecolonies/blob/1c86e4ce5bb6d3ca11b572017dcfd315270f4465/src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java#L325): Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [Permissions.java#L786](https://github.com/Minecolonies/minecolonies/blob/1c86e4ce5bb6d3ca11b572017dcfd315270f4465/src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java#L786): Remove this unused private "isSubscriber" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)
1. ![MAJOR][MAJOR] [Permissions.java#L871](https://github.com/Minecolonies/minecolonies/blob/1c86e4ce5bb6d3ca11b572017dcfd315270f4465/src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java#L871): Reduce this class from 116 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [PermissionsMessage.java#L48](https://github.com/Minecolonies/minecolonies/blob/1c86e4ce5bb6d3ca11b572017dcfd315270f4465/src/main/java/com/minecolonies/coremod/network/messages/PermissionsMessage.java#L48): Reduce this class from 33 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [PermissionsMessage.java#L99](https://github.com/Minecolonies/minecolonies/blob/1c86e4ce5bb6d3ca11b572017dcfd315270f4465/src/main/java/com/minecolonies/coremod/network/messages/PermissionsMessage.java#L99): Reduce this class from 63 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [PermissionsMessage.java#L185](https://github.com/Minecolonies/minecolonies/blob/1c86e4ce5bb6d3ca11b572017dcfd315270f4465/src/main/java/com/minecolonies/coremod/network/messages/PermissionsMessage.java#L185): Reduce this class from 43 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [PermissionsMessage.java#L247](https://github.com/Minecolonies/minecolonies/blob/1c86e4ce5bb6d3ca11b572017dcfd315270f4465/src/main/java/com/minecolonies/coremod/network/messages/PermissionsMessage.java#L247): Reduce this class from 59 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [PermissionsMessage.java#L331](https://github.com/Minecolonies/minecolonies/blob/1c86e4ce5bb6d3ca11b572017dcfd315270f4465/src/main/java/com/minecolonies/coremod/network/messages/PermissionsMessage.java#L331): Reduce this class from 46 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MINOR][MINOR] [blockout](https://github.com/Minecolonies/minecolonies/blob/1c86e4ce5bb6d3ca11b572017dcfd315270f4465/src/main/java/com/minecolonies/blockout): Add a 'package-info.java' file to document the 'blockout' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by xavierh** [comment]
LGTM

[Approved with PullApprove](https://pullapprove.com/Minecolonies/minecolonies/pull-request/1089/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

**Comment by marvin-bitterlich** [comment]
LGTM

[Approved with PullApprove](https://pullapprove.com/Minecolonies/minecolonies/pull-request/1089/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

**Pull request merged**